consensus/storage_fault_monitor_test.go

```diff
@@ -0,0 +1,144 @@
+package consensus_test
+
+import (
+	"context"
+	"errors"
+	"testing"
+
+	"github.com/ipfs/go-cid"
+	cbor "github.com/ipfs/go-ipld-cbor"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/filecoin-project/go-filecoin/abi"
+	"github.com/filecoin-project/go-filecoin/actor/builtin/miner"
+	"github.com/filecoin-project/go-filecoin/address"
+	. "github.com/filecoin-project/go-filecoin/consensus"
+	tf "github.com/filecoin-project/go-filecoin/testhelpers/testflags"
+	"github.com/filecoin-project/go-filecoin/types"
+)
+
+func TestStorageFaultSlasher_Slash(t *testing.T) {
+	tf.UnitTest(t)
+
+	ctx := context.Background()
+	signer, _ := types.NewMockSignersAndKeyInfo(1)
+
+	t.Run("When no miners, empty LateMiners", func(t *testing.T) {
+
+		height := types.NewBlockHeight(1)
+		data, err := cbor.DumpObject(&map[string]uint64{})
+		require.NoError(t, err)
+		queryer := makeQueryer([][]byte{data})
+		minerOwnerAddr := signer.Addresses[0]
+
+		ob := outbox{}
+		fm := NewStorageFaultMonitor(&storageFaultMonitorPorcelain{false, false, queryer}, &ob, minerOwnerAddr)
+		err = fm.Slash(ctx, height)
+		require.NoError(t, err)
+	})
+
+	t.Run("When 3 late miners, sends 3 messages", func(t *testing.T) {
+		getf := address.NewForTestGetter()
+		height := types.NewBlockHeight(100)
+
+		addr1 := getf().String()
+		addr2 := getf().String()
+		addr3 := getf().String()
+
+		data, err := cbor.DumpObject(&map[string]uint64{
+			addr1: miner.PoStStateAfterProvingPeriod,
+			addr2: miner.PoStStateAfterGenerationAttackThreshold,
+			addr3: miner.PoStStateAfterProvingPeriod,
+		})
+		require.NoError(t, err)
+
+		queryer := makeQueryer([][]byte{data})
+		ob := outbox{}
+		minerOwnerAddr := signer.Addresses[0]
+		fm := NewStorageFaultMonitor(&storageFaultMonitorPorcelain{false, false, queryer}, &ob, minerOwnerAddr)
+		err = fm.Slash(ctx, height)
+		assert.NoError(t, err)
+		assert.Equal(t, 3, ob.msgCount)
+	})
+
+	t.Run("When Send fails, get collection of all errors + no messages", func(t *testing.T) {
+		getf := address.NewForTestGetter()
+		height := types.NewBlockHeight(100)
+		addr1 := getf().String()
+		addr2 := getf().String()
+
+		data, err := cbor.DumpObject(&map[string]uint64{
+			addr1: miner.PoStStateAfterProvingPeriod,
+			addr2: miner.PoStStateAfterProvingPeriod,
+		})
+		require.NoError(t, err)
+
+		queryer := makeQueryer([][]byte{data})
+		ob := outbox{failSend: true, failErr: "Boom"}
+		minerOwnerAddr := signer.Addresses[0]
+		fm := NewStorageFaultMonitor(&storageFaultMonitorPorcelain{false, false, queryer}, &ob, minerOwnerAddr)
+		err = fm.Slash(ctx, height)
+		assert.Error(t, err, "Boom\nBoom")
```

Comment:
This assertion is misleading. It's just testing that the error is non nil (and it'll message "Boom\nBoom" if it is nil. Better to `asserrt.Error(t, err)` and then `assert.Contains(t, "message", err.Error()`.